Add camera to iiwa + wsg simulation.

EricCousineau-TRI opened the pull request:
This adds a simple usage of `RgbdCamera` with an Asus Xtion (kindly modeled by Alex) in a modular fashion (but specific to Iiwa + WSG).

This can be further generalized into more components, but the main goal is to have a simple drop-in interface with Naveen's push+pick development branch.

\cc @naveenoid @kunimatsu-tri

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/6906)
<!-- Reviewable:end -->

EricCousineau-TRI commented:
+@kunimatsu-tri for feature review, please.

---

Review status: 0 of 11 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6906#-:-KsKPTjgaF8OmJhvsDG_:b-1actl4)*
<!-- Sent from Reviewable.io -->

kunimatsu-tri commented:
Reviewed 8 of 11 files at r1, 1 of 3 files at r2.
Review status: 7 of 11 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[drake/examples/kuka_iiwa_arm/iiwa_camera.cc, line 96 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6906#-KsKg1vOH3n4iiETNCER:-KsKg1vOH3n4iiETNCES:bgxunx5) ([raw file](https://github.com/robotlocomotion/drake/blob/eb65553f88f1a9ce1415837c8dca77d6d3024b18/drake/examples/kuka_iiwa_arm/iiwa_camera.cc#L96)):*
> ```cpp
> 
>   auto* rgbd_camera = builder.AddSystem<RgbdCamera>(
>       name_, *tree_, *sensor_frame_, M_PI_4, true);
> ```

Recently I've changed `RgbdCamera`'s contractor so that it takes `depth_range_near` and `depth_range_far`.  You can use default arguments here. 
For more details, see: https://github.com/RobotLocomotion/drake/blob/master/drake/systems/sensors/rgbd_camera.h#L181

---

*[drake/examples/kuka_iiwa_arm/iiwa_camera.h, line 32 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6906#-KsKYGjYhMMvi9k_7MoD:-KsKYGjZlNrF2iUzqFV4:b-vkdu1i) ([raw file](https://github.com/robotlocomotion/drake/blob/37d6d1ff269f55f1454d457600a95753fedb9b45/drake/examples/kuka_iiwa_arm/iiwa_camera.h#L32)):*
> ```objc
> // TODO(eric.cousineau): Encode these in an URDF somewhere, where they are
> // easily accessible? (and easily interpreted?)
> class IiwaCameraFrames {
> ```

[Google C++ Style Guide](https://google.github.io/styleguide/cppguide.html#Nonmember,_Static_Member,_and_Global_Functions) says: "Prefer grouping functions with a namespace instead of using a class as if it were a namespace."

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6906)*
<!-- Sent from Reviewable.io -->

EricCousineau-TRI commented:
I've folded in the generic `RgbdCameraDiscrete` changes. Please me know if you'd like me to split that off into a separate PR.

---

Review status: 3 of 15 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/examples/kuka_iiwa_arm/iiwa_camera.cc, line 96 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6906#-KsKg1vOH3n4iiETNCER:-KsKs8_Di7YAtig_0wYs:bjss83q) ([raw file](https://github.com/robotlocomotion/drake/blob/eb65553f88f1a9ce1415837c8dca77d6d3024b18/drake/examples/kuka_iiwa_arm/iiwa_camera.cc#L96)):*
<details><summary><i>Previously, kunimatsu-tri (Kunimatsu Hashimoto) wrote…</i></summary><blockquote>

Recently I've changed `RgbdCamera`'s contractor so that it takes `depth_range_near` and `depth_range_far`.  You can use default arguments here. 
For more details, see: https://github.com/RobotLocomotion/drake/blob/master/drake/systems/sensors/rgbd_camera.h#L181
</blockquote></details>

I'd prefer to keep this as a very simple convenience wrapper, specific to the Xtoin, and not expose all arguments.
I've added `RgbdCameraDiscrete` for a more generic interface.

---

*[drake/examples/kuka_iiwa_arm/iiwa_camera.h, line 32 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6906#-KsKYGjYhMMvi9k_7MoD:-KsKsEwGKwkty8dYli-G:befyqlu) ([raw file](https://github.com/robotlocomotion/drake/blob/37d6d1ff269f55f1454d457600a95753fedb9b45/drake/examples/kuka_iiwa_arm/iiwa_camera.h#L32)):*
<details><summary><i>Previously, kunimatsu-tri (Kunimatsu Hashimoto) wrote…</i></summary><blockquote>

[Google C++ Style Guide](https://google.github.io/styleguide/cppguide.html#Nonmember,_Static_Member,_and_Global_Functions) says: "Prefer grouping functions with a namespace instead of using a class as if it were a namespace."
</blockquote></details>

It feels a tad odd as these are very topical, and more-or-less constant-related (with more involved expressions). As an alternative, I have hidden them.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6906#-:-KsKvnkP7nZlU0Txco82:b34mzd7)*
<!-- Sent from Reviewable.io -->

kunimatsu-tri commented:
Yes, I think it would be better to have separate PR for `RgbdCameraDiscrete` stuff to provide clearer scope for platform reviewer. Could you do that?

---

Reviewed 1 of 6 files at r2, 2 of 8 files at r3.
Review status: 5 of 15 files reviewed at latest revision, 7 unresolved discussions.

---

*[drake/examples/kuka_iiwa_arm/iiwa_xtion.cc, line 96 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6906#-KsKg1vOH3n4iiETNCER:-KsLM0O7fpax5n7Hiydn:b-9rft7l) ([raw file](https://github.com/robotlocomotion/drake/blob/eb65553f88f1a9ce1415837c8dca77d6d3024b18/drake/examples/kuka_iiwa_arm/iiwa_xtion.cc#L96)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

I'd prefer to keep this as a very simple convenience wrapper, specific to the Xtoin, and not expose all arguments.
I've added `RgbdCameraDiscrete` for a more generic interface.
</blockquote></details>

I meant that you're setting `M_PI_4` for `depth_range_near` and `true` for `depth_range_far` which are not what you intended.

---

*[drake/examples/kuka_iiwa_arm/iiwa_xtion.cc, line 71 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/6906#-KsLRkbSnVsLSf1M3EN2:-KsLRkbSnVsLSf1M3EN3:b-71kgaj) ([raw file](https://github.com/robotlocomotion/drake/blob/29e3e75a9dcf1f8d79e84bc1a8f6969c52c16993/drake/examples/kuka_iiwa_arm/iiwa_xtion.cc#L71)):*
> ```cpp
>     : tree_(&tree_builder->tree()) {
>   std::string urdf_resource =
>       "drake/examples/kuka_iiwa_arm/models/cameras/xtion.urdf";
> ```

`xtion.urdf` model should also be placed in somewhere like `drake/systems/sensors/models`. See the comment below.

---

*[drake/examples/kuka_iiwa_arm/iiwa_xtion.h, line 20 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/6906#-KsLR95uGRrjuz6Vm6oi:-KsLR95uGRrjuz6Vm6oj:brrf3ib) ([raw file](https://github.com/robotlocomotion/drake/blob/29e3e75a9dcf1f8d79e84bc1a8f6969c52c16993/drake/examples/kuka_iiwa_arm/iiwa_xtion.h#L20)):*
> ```objc
> 
> /**
>  * Attaches an Asus Xtion PRO camera to a given frame on a RigidBodyTree
> ```

It should be `Xtion PRO LIVE` instead of `Xtion PRO`. The latter is depth only version, no rgb.

---

*[drake/examples/kuka_iiwa_arm/iiwa_xtion.h, line 45 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/6906#-KsLNlop373aX5_hA637:-KsLNlop373aX5_hA638:b-41dr3w) ([raw file](https://github.com/robotlocomotion/drake/blob/29e3e75a9dcf1f8d79e84bc1a8f6969c52c16993/drake/examples/kuka_iiwa_arm/iiwa_xtion.h#L45)):*
> ```objc
>  * @endcode
>  */
> class IiwaXtion : public systems::Diagram<double> {
> ```

I think this class can be more general.  In other words, this can be just `Xtion` instead of `IiwaXtion`. 

And if you agree, the right place to put the code is probably `drake/systems/sensors`.

---

*[drake/systems/sensors/rgbd_camera.h, line 6 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6906#-KsKtQolm0MLNgUjTM4i:-KsKtQolm0MLNgUjTM4j:b6aw2ta) ([raw file](https://github.com/robotlocomotion/drake/blob/f6f5edd0e76c24061322a004890a72f05cbe52ef/drake/systems/sensors/rgbd_camera.h#L6)):*
> ```objc
> #include <memory>
> #include <string>
> #include <utility>
> ```

BTW what is this for?

---

*[drake/systems/sensors/rgbd_camera_publish_lcm_example.cc, line 74 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6906#-KsKtrg1kvFGIOuCWIKs:-KsKtrg1kvFGIOuCWIKt:b-f7wis0) ([raw file](https://github.com/robotlocomotion/drake/blob/f6f5edd0e76c24061322a004890a72f05cbe52ef/drake/systems/sensors/rgbd_camera_publish_lcm_example.cc#L74)):*
> ```cpp
>               "rgbd_camera", plant->get_rigid_body_tree(),
>               Eigen::Vector3d(-1., 0., 1.),
>               Eigen::Vector3d(0., M_PI_4, 0.), M_PI_4, true),
> ```

This is not your fault, but could you remove the last two arguments for `RgbdCamera`, `M_PI_4` and `true`? 
Thanks!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6906#-:-KsLMIYsgoQUsp0S4b4a:b-7j7ngo)*
<!-- Sent from Reviewable.io -->

kunimatsu-tri commented:
Reviewed 1 of 15 files at r1, 2 of 8 files at r3, 11 of 11 files at r4.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[drake/systems/sensors/BUILD, line 326 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/6906#-KsLUrcPRmKqGDZcuIwZ:-KsLUrcPRmKqGDZcuIw_:b-f0c33m) ([raw file](https://github.com/robotlocomotion/drake/blob/55ddddb927f544e44d0afbb390fd9fdccc10bebf/drake/systems/sensors/BUILD#L326)):*
> ```Python
> )
> 
> # TODO(jamiesnape): Remove the manual tag when Xvfb is available on CI.
> ```

This TODO shouldn't be removed yet.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6906)*
<!-- Sent from Reviewable.io -->

EricCousineau-TRI commented:
I have submitted PR #6942 to divvy things up. I will place this PR on hold until that one merges.

---

Reviewed 4 of 18 files at r1, 4 of 9 files at r3, 11 of 14 files at r4.
Review status: 4 of 20 files reviewed at latest revision, 7 unresolved discussions.

---

*[drake/manipulation/sensors/xtion.cc, line 96 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6906#-KsKg1vOH3n4iiETNCER:-KsowSXr6xElR8-WR-98:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/eb65553f88f1a9ce1415837c8dca77d6d3024b18/drake/manipulation/sensors/xtion.cc#L96)):*
<details><summary><i>Previously, kunimatsu-tri (Kunimatsu Hashimoto) wrote…</i></summary><blockquote>

I meant that you're setting `M_PI_4` for `depth_range_near` and `true` for `depth_range_far` which are not what you intended.
</blockquote></details>

Done.

---

*[drake/manipulation/sensors/xtion.cc, line 71 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/6906#-KsLRkbSnVsLSf1M3EN2:-KsowRaGv-MStfj7AXHq:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/29e3e75a9dcf1f8d79e84bc1a8f6969c52c16993/drake/manipulation/sensors/xtion.cc#L71)):*
<details><summary><i>Previously, kunimatsu-tri (Kunimatsu Hashimoto) wrote…</i></summary><blockquote>

`xtion.urdf` model should also be placed in somewhere like `drake/systems/sensors/models`. See the comment below.
</blockquote></details>

Done.

---

*[drake/systems/sensors/BUILD, line 326 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/6906#-KsLUrcPRmKqGDZcuIwZ:-KsowUBLIS1wDsenOo54:bo52puj) ([raw file](https://github.com/robotlocomotion/drake/blob/55ddddb927f544e44d0afbb390fd9fdccc10bebf/drake/systems/sensors/BUILD#L326)):*
<details><summary><i>Previously, kunimatsu-tri (Kunimatsu Hashimoto) wrote…</i></summary><blockquote>

This TODO shouldn't be removed yet.
</blockquote></details>

Done. (This was actually just moved down a few lines.)

---

*[drake/systems/sensors/rgbd_camera.h, line 6 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6906#-KsKtQolm0MLNgUjTM4i:-Ksow_W5gxgqe8WSV_8A:b1whevk) ([raw file](https://github.com/robotlocomotion/drake/blob/f6f5edd0e76c24061322a004890a72f05cbe52ef/drake/systems/sensors/rgbd_camera.h#L6)):*
<details><summary><i>Previously, kunimatsu-tri (Kunimatsu Hashimoto) wrote…</i></summary><blockquote>

BTW what is this for?
</blockquote></details>

BTW Sorry, that was supposed to go into `*.cc`. Fixed in separate PR. (This PR has been rebased against it.)

---

*[drake/systems/sensors/rgbd_camera_publish_lcm_example.cc, line 74 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6906#-KsKtrg1kvFGIOuCWIKs:-KsowiGkLnR5C_0sKnKE:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/f6f5edd0e76c24061322a004890a72f05cbe52ef/drake/systems/sensors/rgbd_camera_publish_lcm_example.cc#L74)):*
<details><summary><i>Previously, kunimatsu-tri (Kunimatsu Hashimoto) wrote…</i></summary><blockquote>

This is not your fault, but could you remove the last two arguments for `RgbdCamera`, `M_PI_4` and `true`? 
Thanks!
</blockquote></details>

Done.

---

*[drake/examples/kuka_iiwa_arm/iiwa_xtion.h, line 20 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/6906#-KsLR95uGRrjuz6Vm6oi:-Ksowkq67JpqwQ-FnnJc:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/29e3e75a9dcf1f8d79e84bc1a8f6969c52c16993/drake/examples/kuka_iiwa_arm/iiwa_xtion.h#L20)):*
<details><summary><i>Previously, kunimatsu-tri (Kunimatsu Hashimoto) wrote…</i></summary><blockquote>

It should be `Xtion PRO LIVE` instead of `Xtion PRO`. The latter is depth only version, no rgb.
</blockquote></details>

Done.

---

*[drake/examples/kuka_iiwa_arm/iiwa_xtion.h, line 45 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/6906#-KsLNlop373aX5_hA637:-KsowkSC1IS0UzKFRFTu:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/29e3e75a9dcf1f8d79e84bc1a8f6969c52c16993/drake/examples/kuka_iiwa_arm/iiwa_xtion.h#L45)):*
<details><summary><i>Previously, kunimatsu-tri (Kunimatsu Hashimoto) wrote…</i></summary><blockquote>

I think this class can be more general.  In other words, this can be just `Xtion` instead of `IiwaXtion`. 

And if you agree, the right place to put the code is probably `drake/systems/sensors`.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6906#-:-KsobRKVvjIjtMS5F-__:b3dlahq)*
<!-- Sent from Reviewable.io -->

EricCousineau-TRI commented:
Resuming progress on this PR, since #6942 has merged.

-@kunimatsu-tri since he is on vacation until Thursday. (Hope you're enjoying it!)
+@naveenoid for feature review please (if you don't have too much else on your review plate)

The main thing that feels awkward in this setup is modifying the RigidBodyTree in the constructor (requiring the odd double-pointer shindig), but still requiring `Build()` to be called.
I think it may fell a little less awkward if perhaps the class does nothing on construction, but then it's usage look like:

```c++
Xtion* camera = new Xtion();
camera->BuildTree(&builder, ...);
...
camera->BuildDiagram(&lcm, ...);
```

Let me know what you think.

---

Review status: 3 of 15 files reviewed at latest revision, 6 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6906#-:-KtEwyCKCrFhSVtgMt_J:bd19jgv)*
<!-- Sent from Reviewable.io -->

EricCousineau-TRI commented:
Review status: 3 of 15 files reviewed at latest revision, 7 unresolved discussions.

---

*[drake/manipulation/models/xtion_description/model.config, line 1 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/6906#-KtExxkoGhK2uGI7fGTM:-KtExxkoGhK2uGI7fGTN:b-9g68o7) ([raw file](https://github.com/robotlocomotion/drake/blob/97dd946cc51ad778843c4bd147f27371634da0b4/drake/manipulation/models/xtion_description/model.config#L1)):*
> ```
> <?xml version="1.0" ?>
> ```

BTW I am doing this to mirror the other models.

@stonier, do you have any suggestions for this to look a little less odd?
(Or should that be slated until we figure out ROS / Gazebo integration more holistically?)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6906)*
<!-- Sent from Reviewable.io -->

naveenoid commented:
First pass done. 

---

Reviewed 1 of 18 files at r1, 2 of 14 files at r4, 11 of 21 files at r6.
Review status: 12 of 15 files reviewed at latest revision, 17 unresolved discussions, some commit checks failed.

---

*[drake/examples/kuka_iiwa_arm/dev/monolithic_pick_and_place/monolithic_pick_and_place_demo.cc, line 157 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/6906#-KtO3eGhotujOx54UnFX:-KtO3eGhotujOx54UnFY:bmo1cxf) ([raw file](https://github.com/robotlocomotion/drake/blob/97dd946cc51ad778843c4bd147f27371634da0b4/drake/examples/kuka_iiwa_arm/dev/monolithic_pick_and_place/monolithic_pick_and_place_demo.cc#L157)):*
> ```cpp
> 
>   if (pcamera) {
>     // Attach Xtion (X) to WSG's end effector (G).
> ```

Why to WSG' end effector? Dont you just mean WSG's body?

---

*[drake/manipulation/models/xtion_description/model.config, line 1 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/6906#-KtO3y4e5iGBUOAKzmHA:-KtO3y4e5iGBUOAKzmHB:b-5nr2ey) ([raw file](https://github.com/robotlocomotion/drake/blob/97dd946cc51ad778843c4bd147f27371634da0b4/drake/manipulation/models/xtion_description/model.config#L1)):*
> ```
> <?xml version="1.0" ?>
> ```

BTW: Just to 100% sure; Are we clear on the licensing front with this?

---

*[drake/manipulation/models/xtion_description/urdf/xtion.urdf, line 8 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/6906#-KtO49g9oCm7rJ4baRTO:-KtO49g9oCm7rJ4baRTP:bqnch1n) ([raw file](https://github.com/robotlocomotion/drake/blob/97dd946cc51ad778843c4bd147f27371634da0b4/drake/manipulation/models/xtion_description/urdf/xtion.urdf#L8)):*
> ```XML
>       <!-- TODO(eric.cousineau): Use actual inertial properties. -->
>       <mass value="0.1"/>
>       <inertia ixx="0.01" ixy="0.0" ixz="0.0" iyy="0.01" iyz="0.0" izz="0.01"/>
> ```

This is easy enough to add already in this PR ( Refer to https://en.wikipedia.org/wiki/List_of_moments_of_inertia#List_of_3D_inertia_tensors for instance)

---

*[drake/manipulation/sensors/BUILD, line 8 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/6906#-KtO4lzm35IOsyBWnaol:-KtO4lzm35IOsyBWnaom:b-xce6z1) ([raw file](https://github.com/robotlocomotion/drake/blob/97dd946cc51ad778843c4bd147f27371634da0b4/drake/manipulation/sensors/BUILD#L8)):*
> ```Python
> # This is tested by usage in:
> # //drake/examples/kuka_iiwa_arm/dev/monolithic_pick_and_place
> #   :monolithic_pick_and_place_demo.
> ```

BTW: I would suggest a TODO here for a future unit test. This is messy.

---

*[drake/manipulation/sensors/xtion.cc, line 31 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/6906#-KtO58ECPffkYOapOq02:-KtO58ECPffkYOapOq03:b-c62f8u) ([raw file](https://github.com/robotlocomotion/drake/blob/97dd946cc51ad778843c4bd147f27371634da0b4/drake/manipulation/sensors/xtion.cc#L31)):*
> ```cpp
> 
> // TODO(eric.cousineau): Encode these in an URDF somewhere, where they are
> // easily accessible? (and easily interpreted?)
> ```

Cant this already be done in this PR? Add an additional link to the Xtion and attach camera to that link.

---

*[drake/manipulation/sensors/xtion.cc, line 83 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/6906#-KtO5OajMtRwXjilONDt:-KtO5OajMtRwXjilONDu:b-9nbhmo) ([raw file](https://github.com/robotlocomotion/drake/blob/97dd946cc51ad778843c4bd147f27371634da0b4/drake/manipulation/sensors/xtion.cc#L83)):*
> ```cpp
>       tree_->get_num_positions() + tree_->get_num_velocities();
> 
>   // @ref https://www.asus.com/us/3D-Sensor/Xtion_PRO_LIVE/specifications/
> ```

These params seem important and a .cc is a pretty hidden place for them. Perhaps a TODO to find a better home for these?

---

*[drake/manipulation/sensors/xtion.h, line 21 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/6906#-KtO5l8WoDIoHzqpSNrR:-KtO5l8WoDIoHzqpSNrS:b7lfk8i) ([raw file](https://github.com/robotlocomotion/drake/blob/97dd946cc51ad778843c4bd147f27371634da0b4/drake/manipulation/sensors/xtion.h#L21)):*
> ```objc
> /**
>  * Attaches an Asus Xtion PRO Live camera to a given frame on a RigidBodyTree
>  * scene, and provides the ability to add the camera as a system in a diagram
> ```

"on a RigidBodyTree scene" -> "on a RigidBodyTree"

---

*[drake/manipulation/util/world_sim_tree_builder.h, line 107 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/6906#-KtO6Dagys0wF04e36yN:-KtO6Dagys0wF04e36yO:b-1vx4bf) ([raw file](https://github.com/robotlocomotion/drake/blob/97dd946cc51ad778843c4bd147f27371634da0b4/drake/manipulation/util/world_sim_tree_builder.h#L107)):*
> ```objc
> 
>   /// Return the (not yet built) mutable tree. Build() must not have been
>   /// called yet.
> ```

BTW: How about stating that "Build() must not have been called yet or else execution halts with an error" or something?

---

*[drake/manipulation/util/world_sim_tree_builder.h, line 112 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/6906#-KtO6hz2hMwpI2dp2Y6H:-KtO6hz2hMwpI2dp2Y6I:b5e84gw) ([raw file](https://github.com/robotlocomotion/drake/blob/97dd946cc51ad778843c4bd147f27371634da0b4/drake/manipulation/util/world_sim_tree_builder.h#L112)):*
> ```objc
>     return *rigid_body_tree_;
>   }
> ```

BTW: I dont really think this mod is needed because you seem to be doing this in order to attach a camera on the object. The workaround with adding a second link at the exact camera location will eliminate the need for this. 

---

*[drake/examples/kuka_iiwa_arm/iiwa_camera.cc, line 99 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6906#-KsKiaScgZJF6RRVo4l6:-KsKiaScgZJF6RRVo4l7:b-cjp7dv) ([raw file](https://github.com/robotlocomotion/drake/blob/eb65553f88f1a9ce1415837c8dca77d6d3024b18/drake/examples/kuka_iiwa_arm/iiwa_camera.cc#L99)):*
> ```cpp
> 
>   // TODO(eric.cousineau): See if there is a better way to connect an input
>   // to other blocks.
> ```

BTW: what do you mean with this comment? 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6906#-:-KtO6Pqbc_jOMdxZOUAV:bc8imr2)*
<!-- Sent from Reviewable.io -->

naveenoid commented:
Reviewed 1 of 14 files at r4, 4 of 21 files at r6.
Review status: all files reviewed at latest revision, 17 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6906)*
<!-- Sent from Reviewable.io -->

EricCousineau-TRI commented:
Review status: 5 of 14 files reviewed at latest revision, 17 unresolved discussions.

---

*[drake/examples/kuka_iiwa_arm/dev/monolithic_pick_and_place/monolithic_pick_and_place_demo.cc, line 157 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/6906#-KtO3eGhotujOx54UnFX:-Kw2eotfKGBW52nVQ0oW:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/97dd946cc51ad778843c4bd147f27371634da0b4/drake/examples/kuka_iiwa_arm/dev/monolithic_pick_and_place/monolithic_pick_and_place_demo.cc#L157)):*
<details><summary><i>Previously, naveenoid (Naveen Kuppuswamy) wrote…</i></summary><blockquote>

Why to WSG' end effector? Dont you just mean WSG's body?
</blockquote></details>

Done.

---

*[drake/manipulation/models/xtion_description/model.config, line 1 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/6906#-KtO3y4e5iGBUOAKzmHA:-Kw2epundc9NW8nH4djK:b-nrtjsq) ([raw file](https://github.com/robotlocomotion/drake/blob/97dd946cc51ad778843c4bd147f27371634da0b4/drake/manipulation/models/xtion_description/model.config#L1)):*
<details><summary><i>Previously, naveenoid (Naveen Kuppuswamy) wrote…</i></summary><blockquote>

BTW: Just to 100% sure; Are we clear on the licensing front with this?
</blockquote></details>

Done. Yup!

---

*[drake/manipulation/models/xtion_description/urdf/xtion.urdf, line 8 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/6906#-KtO49g9oCm7rJ4baRTO:-Kw2ez-LxwPqs7zLPuN6:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/97dd946cc51ad778843c4bd147f27371634da0b4/drake/manipulation/models/xtion_description/urdf/xtion.urdf#L8)):*
<details><summary><i>Previously, naveenoid (Naveen Kuppuswamy) wrote…</i></summary><blockquote>

This is easy enough to add already in this PR ( Refer to https://en.wikipedia.org/wiki/List_of_moments_of_inertia#List_of_3D_inertia_tensors for instance)
</blockquote></details>

Done.

---

*[drake/manipulation/sensors/BUILD, line 8 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/6906#-KtO4lzm35IOsyBWnaol:-Kw2f1Hl6RlA4XJt0dgY:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/97dd946cc51ad778843c4bd147f27371634da0b4/drake/manipulation/sensors/BUILD#L8)):*
<details><summary><i>Previously, naveenoid (Naveen Kuppuswamy) wrote…</i></summary><blockquote>

BTW: I would suggest a TODO here for a future unit test. This is messy.
</blockquote></details>

Done.

---

*[drake/manipulation/sensors/xtion.cc, line 96 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6906#-KsKg1vOH3n4iiETNCER:-Kw2f6tqPPH7OO-WSJnr:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/eb65553f88f1a9ce1415837c8dca77d6d3024b18/drake/manipulation/sensors/xtion.cc#L96)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Done.

---

*[drake/manipulation/sensors/xtion.cc, line 99 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6906#-KsKiaScgZJF6RRVo4l6:-Kw2f8GtAZTRyC_WzScn:bp3bkxe) ([raw file](https://github.com/robotlocomotion/drake/blob/eb65553f88f1a9ce1415837c8dca77d6d3024b18/drake/manipulation/sensors/xtion.cc#L99)):*
<details><summary><i>Previously, naveenoid (Naveen Kuppuswamy) wrote…</i></summary><blockquote>

BTW: what do you mean with this comment? 
</blockquote></details>

Reworded to clarify.

---

*[drake/manipulation/sensors/xtion.cc, line 71 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/6906#-KsLRkbSnVsLSf1M3EN2:-Kw2f52_f7ZudN_2_WeL:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/29e3e75a9dcf1f8d79e84bc1a8f6969c52c16993/drake/manipulation/sensors/xtion.cc#L71)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Done.

---

*[drake/manipulation/sensors/xtion.cc, line 31 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/6906#-KtO58ECPffkYOapOq02:-Kw2f377E0FhUTHKhar0:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/97dd946cc51ad778843c4bd147f27371634da0b4/drake/manipulation/sensors/xtion.cc#L31)):*
<details><summary><i>Previously, naveenoid (Naveen Kuppuswamy) wrote…</i></summary><blockquote>

Cant this already be done in this PR? Add an additional link to the Xtion and attach camera to that link.
</blockquote></details>

Done.

---

*[drake/manipulation/sensors/xtion.cc, line 83 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/6906#-KtO5OajMtRwXjilONDt:-Kw2f65m9cRv5s_u5E7i:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/97dd946cc51ad778843c4bd147f27371634da0b4/drake/manipulation/sensors/xtion.cc#L83)):*
<details><summary><i>Previously, naveenoid (Naveen Kuppuswamy) wrote…</i></summary><blockquote>

These params seem important and a .cc is a pretty hidden place for them. Perhaps a TODO to find a better home for these?
</blockquote></details>

Done.

---

*[drake/manipulation/sensors/xtion.h, line 21 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/6906#-KtO5l8WoDIoHzqpSNrR:-Kw2fDGvE9ALnTDqOnaI:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/97dd946cc51ad778843c4bd147f27371634da0b4/drake/manipulation/sensors/xtion.h#L21)):*
<details><summary><i>Previously, naveenoid (Naveen Kuppuswamy) wrote…</i></summary><blockquote>

"on a RigidBodyTree scene" -> "on a RigidBodyTree"
</blockquote></details>

Done.

---

*[drake/manipulation/util/world_sim_tree_builder.h, line 107 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/6906#-KtO6Dagys0wF04e36yN:-Kw2fF7H6pfnxn97LiOd:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/97dd946cc51ad778843c4bd147f27371634da0b4/drake/manipulation/util/world_sim_tree_builder.h#L107)):*
<details><summary><i>Previously, naveenoid (Naveen Kuppuswamy) wrote…</i></summary><blockquote>

BTW: How about stating that "Build() must not have been called yet or else execution halts with an error" or something?
</blockquote></details>

Done.

---

*[drake/manipulation/util/world_sim_tree_builder.h, line 112 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/6906#-KtO6hz2hMwpI2dp2Y6H:-Kw2gf_96pXL_VSzhkx2:b-4ak7cx) ([raw file](https://github.com/robotlocomotion/drake/blob/97dd946cc51ad778843c4bd147f27371634da0b4/drake/manipulation/util/world_sim_tree_builder.h#L112)):*
<details><summary><i>Previously, naveenoid (Naveen Kuppuswamy) wrote…</i></summary><blockquote>

BTW: I dont really think this mod is needed because you seem to be doing this in order to attach a camera on the object. The workaround with adding a second link at the exact camera location will eliminate the need for this. 
</blockquote></details>

True; but for now, I'd like to keep this interface simple, such that if a person wants a fixed-frame instance, they will simply add a frame the rigid body tree, rather than delegate to the alternative `RgbdCamera` constructor (which would require propagating a few more items here).

Ultimately, it'd be nice to not even need this class, but rather have sensor tags, so I'd prefer not to go overboard with purity here.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6906)*
<!-- Sent from Reviewable.io -->

stonier commented:
*[drake/manipulation/models/xtion_description/model.config, line 1 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/6906#-KtExxkoGhK2uGI7fGTM:-Kw5YDW2h2LHREw0uqGT:bypp3j4) ([raw file](https://github.com/robotlocomotion/drake/blob/97dd946cc51ad778843c4bd147f27371634da0b4/drake/manipulation/models/xtion_description/model.config#L1)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

BTW I am doing this to mirror the other models.

@stonier, do you have any suggestions for this to look a little less odd?
(Or should that be slated until we figure out ROS / Gazebo integration more holistically?)
</blockquote></details>

If you don't have a strong reason motivating it to be done in a different way, emulating what is already familiar to people is probably a good thing to do.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6906)*
<!-- Sent from Reviewable.io -->

EricCousineau-TRI commented:
Review status: 5 of 14 files reviewed at latest revision, 8 unresolved discussions.

---

*[drake/manipulation/util/world_sim_tree_builder.h, line 112 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/6906#-KtO6hz2hMwpI2dp2Y6H:-Kw5i-tx35dRn6gHEHm4:b-yd12lb) ([raw file](https://github.com/robotlocomotion/drake/blob/97dd946cc51ad778843c4bd147f27371634da0b4/drake/manipulation/util/world_sim_tree_builder.h#L112)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

True; but for now, I'd like to keep this interface simple, such that if a person wants a fixed-frame instance, they will simply add a frame the rigid body tree, rather than delegate to the alternative `RgbdCamera` constructor (which would require propagating a few more items here).

Ultimately, it'd be nice to not even need this class, but rather have sensor tags, so I'd prefer not to go overboard with purity here.
</blockquote></details>

(But, if you or @kunimatsu-tri have objections to having a ctor different than RgbdCamera, please let me know and I'll change it, and could possibly remove this method if I don't need to create a world frame.)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6906)*
<!-- Sent from Reviewable.io -->

naveenoid commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 5 of 14 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/manipulation/util/world_sim_tree_builder.h, line 107 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/6906#-KtO6Dagys0wF04e36yN:-Kw6P9yrU0zgBGqY-84_:bf7cigs) ([raw file](https://github.com/robotlocomotion/drake/blob/97dd946cc51ad778843c4bd147f27371634da0b4/drake/manipulation/util/world_sim_tree_builder.h#L107)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

Done.
</blockquote></details>

I dont really see a change here except a reformat but I leave it to you.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6906#-:-Kw6PJSz-KiEhFHFZnLa:bnfp4nl)*
<!-- Sent from Reviewable.io -->

EricCousineau-TRI commented:
Review status: 5 of 14 files reviewed at latest revision, all discussions resolved.

---

*[drake/manipulation/util/world_sim_tree_builder.h, line 107 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/6906#-KtO6Dagys0wF04e36yN:-Kw6Vv0U-VsvYuSssQtI:b-h51c29) ([raw file](https://github.com/robotlocomotion/drake/blob/97dd946cc51ad778843c4bd147f27371634da0b4/drake/manipulation/util/world_sim_tree_builder.h#L107)):*
<details><summary><i>Previously, naveenoid (Naveen Kuppuswamy) wrote…</i></summary><blockquote>

I dont really see a change here except a reformat but I leave it to you.
</blockquote></details>

For me, `@pre` means that this precondition *must* be satisfied, or bad things happen. Could throw or abort, but since it's not a `@throws`, then peeps should assume contract violation implies an abort. Saying execution halts seems redundant.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6906)*
<!-- Sent from Reviewable.io -->

EricCousineau-TRI commented:
+@sammy-tri for platform review, please.

---

Review status: 5 of 14 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6906#-:-Kw6WV3zcinNFZ8bd5j0:bz1t39v)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
Notes added from f2f discussion.  Will finish first pass tomorrow.

---

Reviewed 2 of 18 files at r1, 1 of 14 files at r4, 3 of 21 files at r6, 3 of 12 files at r7.
Review status: 8 of 14 files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/examples/kuka_iiwa_arm/dev/monolithic_pick_and_place/monolithic_pick_and_place_demo.cc, line 172 at r8](https://reviewable.io:443/reviews/robotlocomotion/drake/6906#-Kw7MJ0vG-uNeYNBK15f:-Kw7MJ0vG-uNeYNBK15g:br1mxzp) ([raw file](https://github.com/robotlocomotion/drake/blob/9a489ed31f8bad43d640e05d4b6fbdcc2a0bd31d/drake/examples/kuka_iiwa_arm/dev/monolithic_pick_and_place/monolithic_pick_and_place_demo.cc#L172)):*
> ```cpp
>         std::make_shared<RigidBodyFrame<double>>("xtion_fixture",
>                                                  wsg_body, X_GX);
>     tree_builder->mutable_tree().addFrame(xtion_fixture);
> ```

From f2f If we added a function to `WorldSimTreeBuilder` which took the body which a new frame should be attached to + the transform and it then returned a shared ptr to the new frame (and called `addFrame()` on it) we could get rid of `mutable_tree()`

---

*[drake/manipulation/models/xtion_description/BUILD, line 6 at r8](https://reviewable.io:443/reviews/robotlocomotion/drake/6906#-Kw76V-DP8OrZVLLweZb:-Kw76V-ELUAAnkRfC8lA:b-r6rv4k) ([raw file](https://github.com/robotlocomotion/drake/blob/9a489ed31f8bad43d640e05d4b6fbdcc2a0bd31d/drake/manipulation/models/xtion_description/BUILD#L6)):*
> ```Python
> load(
>     "//tools:drake.bzl",
>     "drake_cc_googletest",
> ```

unused?  If this were a more complex model I would argue there should be some kind of test which loads it and verifies that it loaded correctly...  but considering it doesn't even have any joints, that might be a bit overkill here.

---

*[drake/manipulation/models/xtion_description/package.xml, line 11 at r8](https://reviewable.io:443/reviews/robotlocomotion/drake/6906#-Kw75uBDl6ItdsdHtdh8:-Kw75uBEW2hW-w_kDiy-:b-qoh27s) ([raw file](https://github.com/robotlocomotion/drake/blob/9a489ed31f8bad43d640e05d4b6fbdcc2a0bd31d/drake/manipulation/models/xtion_description/package.xml#L11)):*
> ```XML
>   <maintainer email="drake-users@mit.edu">Drake Users</maintainer>
>   <license>BSD</license>
>   <buildtool_depend>catkin</buildtool_depend>
> ```

This `buildtool_depend` bit doesn't seem accurate to me?  I don't think there are any build steps required for this urdf?  (I also don't know what  I'm talking about when it comes to ROS packages, so apologies if I'm way off here)

---

*[drake/manipulation/sensors/xtion.h, line 59 at r8](https://reviewable.io:443/reviews/robotlocomotion/drake/6906#-Kw7M2mQrflzkcIH1r83:-Kw7M2mQrflzkcIH1r84:b-9ueufd) ([raw file](https://github.com/robotlocomotion/drake/blob/9a489ed31f8bad43d640e05d4b6fbdcc2a0bd31d/drake/manipulation/sensors/xtion.h#L59)):*
> ```objc
>    */
>   Xtion(const std::string& name,
>         util::WorldSimTreeBuilder<double>* tree_builder,
> ```

From f2f:  Propose not using `WorldSimTreeBuilder` at all from this class and having a single constructor which takes the fixture frame and the sensor frame for later use when the diagram is built.  (I'd also be good with a zero-arg constructor and waiting to pass them into Build...   or building in the constructor if that seems feasible.)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6906#-:-Kw7RcXrazdjaryDgN-T:b-s31y15)*
<!-- Sent from Reviewable.io -->

EricCousineau-TRI commented:
Review status: 6 of 17 files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/examples/kuka_iiwa_arm/dev/monolithic_pick_and_place/monolithic_pick_and_place_demo.cc, line 172 at r8](https://reviewable.io:443/reviews/robotlocomotion/drake/6906#-Kw7MJ0vG-uNeYNBK15f:-Kwa_XPUQPbhhwxfJArX:bs62ej4) ([raw file](https://github.com/robotlocomotion/drake/blob/9a489ed31f8bad43d640e05d4b6fbdcc2a0bd31d/drake/examples/kuka_iiwa_arm/dev/monolithic_pick_and_place/monolithic_pick_and_place_demo.cc#L172)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

From f2f If we added a function to `WorldSimTreeBuilder` which took the body which a new frame should be attached to + the transform and it then returned a shared ptr to the new frame (and called `addFrame()` on it) we could get rid of `mutable_tree()`
</blockquote></details>

Done. Took Naveen's advice and placed this into a fixture URDF.

---

*[drake/manipulation/models/xtion_description/BUILD.bazel, line 6 at r8](https://reviewable.io:443/reviews/robotlocomotion/drake/6906#-Kw76V-DP8OrZVLLweZb:-Kwa_4NZxRMohQ6d5mJp:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/9a489ed31f8bad43d640e05d4b6fbdcc2a0bd31d/drake/manipulation/models/xtion_description/BUILD.bazel#L6)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

unused?  If this were a more complex model I would argue there should be some kind of test which loads it and verifies that it loaded correctly...  but considering it doesn't even have any joints, that might be a bit overkill here.
</blockquote></details>

Done.

---

*[drake/manipulation/models/xtion_description/package.xml, line 11 at r8](https://reviewable.io:443/reviews/robotlocomotion/drake/6906#-Kw75uBDl6ItdsdHtdh8:-Kw_kGz6CmQxANoG0Il2:b-c6jjko) ([raw file](https://github.com/robotlocomotion/drake/blob/9a489ed31f8bad43d640e05d4b6fbdcc2a0bd31d/drake/manipulation/models/xtion_description/package.xml#L11)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

This `buildtool_depend` bit doesn't seem accurate to me?  I don't think there are any build steps required for this urdf?  (I also don't know what  I'm talking about when it comes to ROS packages, so apologies if I'm way off here)
</blockquote></details>

While this is indeed not accurate (as there should be a `CMakeLists.txt`), this is copied from `iiwa_description`. I'd prefer to keep it consistent, if possible.

---

*[drake/manipulation/sensors/xtion.h, line 59 at r8](https://reviewable.io:443/reviews/robotlocomotion/drake/6906#-Kw7M2mQrflzkcIH1r83:-Kwa_5VKLTPUCu94KX_n:b-kuec5h) ([raw file](https://github.com/robotlocomotion/drake/blob/9a489ed31f8bad43d640e05d4b6fbdcc2a0bd31d/drake/manipulation/sensors/xtion.h#L59)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

From f2f:  Propose not using `WorldSimTreeBuilder` at all from this class and having a single constructor which takes the fixture frame and the sensor frame for later use when the diagram is built.  (I'd also be good with a zero-arg constructor and waiting to pass them into Build...   or building in the constructor if that seems feasible.)
</blockquote></details>

Per follow-up f2f, in order to enable encapsulation (and make an ease-of-use wrapper), this API still depends on `WorldSimTreeBuilder`, but delegates this operation to a method rather than the constructor.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6906)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
First pass complete.

---

Reviewed 1 of 14 files at r4, 1 of 23 files at r6, 13 of 15 files at r9.
Review status: all files reviewed at latest revision, 11 unresolved discussions.

---

*[drake/examples/kuka_iiwa_arm/dev/monolithic_pick_and_place/monolithic_pick_and_place_demo.cc, line 280 at r9](https://reviewable.io:443/reviews/robotlocomotion/drake/6906#-Kwaxllfvi5uCuagwCnu:-Kwaxllg-yqT7EDONAnu:b-mdx63y) ([raw file](https://github.com/robotlocomotion/drake/blob/5115031b8495c348b80563a65292fdca920bdce2/drake/examples/kuka_iiwa_arm/dev/monolithic_pick_and_place/monolithic_pick_and_place_demo.cc#L280)):*
> ```cpp
>   // TODO(eric.cousineau): Fix this once we can actually publish according to
>   // major timesteps (#5629).
>   const double visualizer_publish_period = 0.001;
> ```

I don't think this TODO adequately explains why we would want to publish the visualizer method at 1kHz.  It almost smells like we're using the visualizer publish period to control some other part of the simulation.

---

*[drake/manipulation/sensors/BUILD.bazel, line 24 at r9](https://reviewable.io:443/reviews/robotlocomotion/drake/6906#-KwaxCHBOzaOYTK8tEao:-KwaxCHCgvTwHFuBR_uu:b-2sgw2u) ([raw file](https://github.com/robotlocomotion/drake/blob/5115031b8495c348b80563a65292fdca920bdce2/drake/manipulation/sensors/BUILD.bazel#L24)):*
> ```
>         "//drake/multibody/rigid_body_plant:frame_visualizer",
>         "//drake/systems/framework:diagram_builder",
>         "//drake/systems/primitives",
> ```

BTW consider listing the two actual primitives used `pass_through` and `zero_order_hold`.

---

*[drake/manipulation/sensors/xtion.cc, line 34 at r9](https://reviewable.io:443/reviews/robotlocomotion/drake/6906#-KwavwV12i5vMJm2Ocyu:-KwavwV2A6kjRYS9hk1c:baz754j) ([raw file](https://github.com/robotlocomotion/drake/blob/5115031b8495c348b80563a65292fdca920bdce2/drake/manipulation/sensors/xtion.cc#L34)):*
> ```cpp
> 
> void Xtion::AddToTree(WorldSimTreeBuilder<double> *tree_builder,
>                     std::shared_ptr<RigidBodyFrame<double>> fixture_frame) {
> ```

odd indentation

---

*[drake/manipulation/sensors/xtion.cc, line 49 at r9](https://reviewable.io:443/reviews/robotlocomotion/drake/6906#-KwavxyD911v2zx7MB2E:-KwavxyEdYg-pzgg70Oj:baz754j) ([raw file](https://github.com/robotlocomotion/drake/blob/5115031b8495c348b80563a65292fdca920bdce2/drake/manipulation/sensors/xtion.cc#L49)):*
> ```cpp
> 
> void Xtion::BuildDiagram(lcm::DrakeLcm* lcm, bool add_lcm_publisher,
>                   bool add_frame_visualizer) {
> ```

odd indentation

---

*[drake/manipulation/sensors/xtion.cc, line 59 at r9](https://reviewable.io:443/reviews/robotlocomotion/drake/6906#-Kwaw-J8sN4c3z6N1-tp:-Kwaw-J8sN4c3z6N1-tq:b-ngufyw) ([raw file](https://github.com/robotlocomotion/drake/blob/5115031b8495c348b80563a65292fdca920bdce2/drake/manipulation/sensors/xtion.cc#L59)):*
> ```cpp
>       tree_->get_num_positions() + tree_->get_num_velocities();
> 
>   // Wrap constants with `double{}` to avoid need for ODR (one definition rule).
> ```

This surprised me that it's necessary for something declared `static constexpr`.  This feels like a bad smell since it implies it's impossible to ever reference these *public* variables directly, which would be a big gotcha if true.  If ODR is really this big of a problem should we do something like declare the variables `static const` in the header without a value and then declare the value in the CC file?  (or if they're not an important part of the public API just declare them in this file only?)

---

*[drake/manipulation/sensors/xtion.cc, line 68 at r9](https://reviewable.io:443/reviews/robotlocomotion/drake/6906#-KwawY_wwbPbVTsaaxps:-KwawY_x24VzfyoKcR5H:b-psvqoj) ([raw file](https://github.com/robotlocomotion/drake/blob/5115031b8495c348b80563a65292fdca920bdce2/drake/manipulation/sensors/xtion.cc#L68)):*
> ```cpp
>   camera->set_name(name_ + "_rgbd_camera");
> 
>   // Use `PassThrough` to connect imported input to multiple inputs within this
> ```

should this say "exported input"?

---

*[drake/manipulation/sensors/xtion.h, line 11 at r9](https://reviewable.io:443/reviews/robotlocomotion/drake/6906#-KwauDjc9m4fpk8J4wsJ:-KwauDjdkysV5Z9I9Ph2:bl0j0fd) ([raw file](https://github.com/robotlocomotion/drake/blob/5115031b8495c348b80563a65292fdca920bdce2/drake/manipulation/sensors/xtion.h#L11)):*
> ```objc
> #include "drake/lcm/drake_lcm.h"
> #include "drake/manipulation/util/world_sim_tree_builder.h"
> #include "drake/multibody/rigid_body_plant/rigid_body_plant.h"
> ```

BTW unused in header consider moving into cc file

---

*[drake/manipulation/sensors/xtion.h, line 12 at r9](https://reviewable.io:443/reviews/robotlocomotion/drake/6906#-KwauGZAmRN3B7CJIFiD:-KwauGZBerSjEneXlfoz:bl0j0fd) ([raw file](https://github.com/robotlocomotion/drake/blob/5115031b8495c348b80563a65292fdca920bdce2/drake/manipulation/sensors/xtion.h#L12)):*
> ```objc
> #include "drake/manipulation/util/world_sim_tree_builder.h"
> #include "drake/multibody/rigid_body_plant/rigid_body_plant.h"
> #include "drake/systems/framework/diagram_builder.h"
> ```

BTW unused in header consider moving into cc file

---

*[drake/manipulation/sensors/xtion.h, line 13 at r9](https://reviewable.io:443/reviews/robotlocomotion/drake/6906#-KwauQG4Rwj812fJKq6f:-KwauQG4Rwj812fJKq6g:bl0j0fd) ([raw file](https://github.com/robotlocomotion/drake/blob/5115031b8495c348b80563a65292fdca920bdce2/drake/manipulation/sensors/xtion.h#L13)):*
> ```objc
> #include "drake/multibody/rigid_body_plant/rigid_body_plant.h"
> #include "drake/systems/framework/diagram_builder.h"
> #include "drake/systems/sensors/rgbd_camera.h"
> ```

BTW unused in header consider moving into cc file

---

*[drake/manipulation/sensors/xtion.h, line 73 at r9](https://reviewable.io:443/reviews/robotlocomotion/drake/6906#-Kwax1oXc-4KtLNg9WvN:-Kwax1oXc-4KtLNg9WvO:b53pf2p) ([raw file](https://github.com/robotlocomotion/drake/blob/5115031b8495c348b80563a65292fdca920bdce2/drake/manipulation/sensors/xtion.h#L73)):*
> ```objc
>   /**
>    * Build diagram containing RGBD camera with Xtion settings.
>    * @param lcm LCM instance. This must not be null of either
> ```

typo 'of' instead of 'if'?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6906#-:-KwayOwwvOU7ew-6VcAW:bvb1a6z)*
<!-- Sent from Reviewable.io -->

EricCousineau-TRI commented:
Address comments - thanks!

---

Reviewed 1 of 14 files at r4, 1 of 23 files at r6, 1 of 13 files at r7, 2 of 15 files at r9, 8 of 8 files at r10.
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---

*[drake/examples/kuka_iiwa_arm/dev/monolithic_pick_and_place/monolithic_pick_and_place_demo.cc, line 280 at r9](https://reviewable.io:443/reviews/robotlocomotion/drake/6906#-Kwaxllfvi5uCuagwCnu:-KyrTjHk8xnv0rAUWC2L:b-1l5k8u) ([raw file](https://github.com/robotlocomotion/drake/blob/5115031b8495c348b80563a65292fdca920bdce2/drake/examples/kuka_iiwa_arm/dev/monolithic_pick_and_place/monolithic_pick_and_place_demo.cc#L280)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

I don't think this TODO adequately explains why we would want to publish the visualizer method at 1kHz.  It almost smells like we're using the visualizer publish period to control some other part of the simulation.
</blockquote></details>

Removed the workaround for now, and left a TODO relevant to dependent issue (#5629).

---

*[drake/manipulation/sensors/BUILD.bazel, line 24 at r9](https://reviewable.io:443/reviews/robotlocomotion/drake/6906#-KwaxCHBOzaOYTK8tEao:-KyrTrY7YmmDrzvZNTG8:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/5115031b8495c348b80563a65292fdca920bdce2/drake/manipulation/sensors/BUILD.bazel#L24)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

BTW consider listing the two actual primitives used `pass_through` and `zero_order_hold`.
</blockquote></details>

Done.

---

*[drake/manipulation/sensors/xtion.cc, line 34 at r9](https://reviewable.io:443/reviews/robotlocomotion/drake/6906#-KwavwV12i5vMJm2Ocyu:-KyrTsqqFlNw4XzGLmAF:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/5115031b8495c348b80563a65292fdca920bdce2/drake/manipulation/sensors/xtion.cc#L34)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

odd indentation
</blockquote></details>

Done.

---

*[drake/manipulation/sensors/xtion.cc, line 49 at r9](https://reviewable.io:443/reviews/robotlocomotion/drake/6906#-KwavxyD911v2zx7MB2E:-KyrTtDwvMkZW5Rurd60:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/5115031b8495c348b80563a65292fdca920bdce2/drake/manipulation/sensors/xtion.cc#L49)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

odd indentation
</blockquote></details>

Done.

---

*[drake/manipulation/sensors/xtion.cc, line 59 at r9](https://reviewable.io:443/reviews/robotlocomotion/drake/6906#-Kwaw-J8sN4c3z6N1-tp:-KyrTzMnRNP4kX6xvhN8:b-9pq3tt) ([raw file](https://github.com/robotlocomotion/drake/blob/5115031b8495c348b80563a65292fdca920bdce2/drake/manipulation/sensors/xtion.cc#L59)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

This surprised me that it's necessary for something declared `static constexpr`.  This feels like a bad smell since it implies it's impossible to ever reference these *public* variables directly, which would be a big gotcha if true.  If ODR is really this big of a problem should we do something like declare the variables `static const` in the header without a value and then declare the value in the CC file?  (or if they're not an important part of the public API just declare them in this file only?)
</blockquote></details>

Done.

TBH I really don't get why `static constexpr` is such a kludgy mess. IIRC, you're stuck between this ODR oddity, or possibly getting multiple definition errors otherwise. No idea why they didn't just say "Oh, just treat it as a literal" and move on...

---

*[drake/manipulation/sensors/xtion.cc, line 68 at r9](https://reviewable.io:443/reviews/robotlocomotion/drake/6906#-KwawY_wwbPbVTsaaxps:-KyrUODrhvtK0w2OGkZ2:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/5115031b8495c348b80563a65292fdca920bdce2/drake/manipulation/sensors/xtion.cc#L68)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

should this say "exported input"?
</blockquote></details>

Done.

---

*[drake/manipulation/sensors/xtion.h, line 11 at r9](https://reviewable.io:443/reviews/robotlocomotion/drake/6906#-KwauDjc9m4fpk8J4wsJ:-KyrUP0UFBLyE31sRtBz:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/5115031b8495c348b80563a65292fdca920bdce2/drake/manipulation/sensors/xtion.h#L11)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

BTW unused in header consider moving into cc file
</blockquote></details>

Done.

---

*[drake/manipulation/sensors/xtion.h, line 12 at r9](https://reviewable.io:443/reviews/robotlocomotion/drake/6906#-KwauGZAmRN3B7CJIFiD:-KyrUPVjCWG9_Lz_xZC-:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/5115031b8495c348b80563a65292fdca920bdce2/drake/manipulation/sensors/xtion.h#L12)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

BTW unused in header consider moving into cc file
</blockquote></details>

Done.

---

*[drake/manipulation/sensors/xtion.h, line 13 at r9](https://reviewable.io:443/reviews/robotlocomotion/drake/6906#-KwauQG4Rwj812fJKq6f:-KyrUPsBdxy92DBw7Rbz:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/5115031b8495c348b80563a65292fdca920bdce2/drake/manipulation/sensors/xtion.h#L13)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

BTW unused in header consider moving into cc file
</blockquote></details>

Done.

---

*[drake/manipulation/sensors/xtion.h, line 73 at r9](https://reviewable.io:443/reviews/robotlocomotion/drake/6906#-Kwax1oXc-4KtLNg9WvN:-KyrUbM87U45zSyPnNgn:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/5115031b8495c348b80563a65292fdca920bdce2/drake/manipulation/sensors/xtion.h#L73)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

typo 'of' instead of 'if'?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6906#-:-KyrUvcAHWIYosb74S93:br1u962)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
Reviewed 1 of 14 files at r4, 1 of 23 files at r6, 2 of 15 files at r9, 8 of 8 files at r10.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/examples/kuka_iiwa_arm/dev/monolithic_pick_and_place/monolithic_pick_and_place_demo.cc, line 280 at r9](https://reviewable.io:443/reviews/robotlocomotion/drake/6906#-Kwaxllfvi5uCuagwCnu:-KyrXW9KdBXYZ0McaLJz:beanx9r) ([raw file](https://github.com/robotlocomotion/drake/blob/5115031b8495c348b80563a65292fdca920bdce2/drake/examples/kuka_iiwa_arm/dev/monolithic_pick_and_place/monolithic_pick_and_place_demo.cc#L280)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

Removed the workaround for now, and left a TODO relevant to dependent issue (#5629).
</blockquote></details>

I wonder if I misunderstood something when I wrote my comment above originallly (I think the answer is that I almost certainly did).  I'm OK with this code as it is now, but if you want to go back to the version which doesn't publish every timestep (and I think I like that better), what about setting the visualizer publish period to FLAGS_dt as the temporary workaround instead of hardcoding 0.001? 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6906)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6906#-:-KyrY4aSAzPXGsWlL5yx:bnfp4nl)*
<!-- Sent from Reviewable.io -->

EricCousineau-TRI commented:
Review status: 16 of 17 files reviewed at latest revision, all discussions resolved.

---

*[drake/examples/kuka_iiwa_arm/dev/monolithic_pick_and_place/monolithic_pick_and_place_demo.cc, line 280 at r9](https://reviewable.io:443/reviews/robotlocomotion/drake/6906#-Kwaxllfvi5uCuagwCnu:-Kyr_S9OxPYxrEnDGu_J:byed8lo) ([raw file](https://github.com/robotlocomotion/drake/blob/5115031b8495c348b80563a65292fdca920bdce2/drake/examples/kuka_iiwa_arm/dev/monolithic_pick_and_place/monolithic_pick_and_place_demo.cc#L280)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

I wonder if I misunderstood something when I wrote my comment above originallly (I think the answer is that I almost certainly did).  I'm OK with this code as it is now, but if you want to go back to the version which doesn't publish every timestep (and I think I like that better), what about setting the visualizer publish period to FLAGS_dt as the temporary workaround instead of hardcoding 0.001? 
</blockquote></details>

Done - thanks!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6906)*
<!-- Sent from Reviewable.io -->

EricCousineau-TRI commented:
Reviewed 2 of 14 files at r4, 3 of 23 files at r6, 2 of 13 files at r7, 3 of 15 files at r9, 1 of 1 files at r11.
Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6906)*
<!-- Sent from Reviewable.io -->
